pkg/serverless/daemon/daemon.go

```diff
@@ -174,49 +218,78 @@ func (d *Daemon) UseAdaptiveFlush(enabled bool) {
 }
 
 // TriggerFlush triggers a flush of the aggregated metrics, traces and logs.
-// They are flushed concurrently.
+// If the flush times out, the daemon will stop waiting for the flush to complete, but the
+// flush may be continued on the next invocation.
 // In some circumstances, it may switch to another flush strategy after the flush.
-// isLastFlush indicates whether this is the last flush before the shutdown or not.
-func (d *Daemon) TriggerFlush(ctx context.Context, isLastFlush bool) {
-	// Increment the invocation wait group which tracks whether work is in progress for the daemon
+func (d *Daemon) TriggerFlush(isLastFlushBeforeShutdown bool) {
 	d.InvcWg.Add(1)
 	defer d.InvcWg.Done()
-	wg := sync.WaitGroup{}
-	wg.Add(1)
-	wg.Add(1)
-	wg.Add(1)
 
-	// metrics
-	go func() {
-		if d.MetricAgent != nil {
-			d.MetricAgent.Flush()
-		}
-		wg.Done()
-	}()
+	ctx, cancel := context.WithTimeout(context.Background(), FlushTimeout)
 
-	// traces
-	go func() {
-		if d.TraceAgent != nil {
-			d.TraceAgent.Get().FlushSync()
-		}
-		wg.Done()
-	}()
+	wg := sync.WaitGroup{}
+	wg.Add(3)
 
-	// logs
-	go func() {
-		logs.Flush(ctx)
-		wg.Done()
-	}()
+	go d.flushMetrics(&wg)
+	go d.flushTraces(&wg)
+	go d.flushLogs(ctx, &wg)
 
-	wg.Wait()
-	log.Debug("Flush done")
+	timedOut := waitWithTimeout(&wg, FlushTimeout)
+	if timedOut {
+		log.Debug("Timed out while flushing, flush may be continued on next invocation")
+	} else {
+		log.Debug("Finished flushing")
+	}
+	cancel()
 
-	// After flushing, re-evaluate flush strategy (if applicable)
-	if !isLastFlush {
+	if !isLastFlushBeforeShutdown {
 		d.UpdateStrategy()
 	}
 }
 
+// flushMetrics flushes aggregated metrics to the intake.
+// It is protected by a mutex to ensure only one metrics flush can be in progress at any given time.
+func (d *Daemon) flushMetrics(wg *sync.WaitGroup) {
+	flushStartTime := time.Now().Unix()
+
+	d.metricsFlushMutex.Lock()
+	log.Debugf("Beginning metrics flush at time %d", flushStartTime)
+	if d.MetricAgent != nil {
+		d.MetricAgent.Flush()
+	}
+	log.Debugf("Finished metrics flush that was started at time %d", flushStartTime)
+	wg.Done()
+	d.metricsFlushMutex.Unlock()
+}
+
+// flushTraces flushes aggregated traces to the intake.
+// It is protected by a mutex to ensure only one traces flush can be in progress at any given time.
+func (d *Daemon) flushTraces(wg *sync.WaitGroup) {
+	flushStartTime := time.Now().Unix()
+
+	d.tracesFlushMutex.Lock()
+	log.Debugf("Beginning traces flush at time %d", flushStartTime)
+	if d.TraceAgent != nil {
+		d.TraceAgent.Get().FlushSync()
```

Comment:
Could you also add .get() != nil ? cc https://github.com/DataDog/datadog-agent/pull/9205